core/src/main/java/io/parsingdata/metal/data/ImmutableList.java

```diff
@@ -59,9 +63,13 @@ private ImmutableList(final T head, final ImmutableList<T> tail) {
 
     public ImmutableList<T> add(final ImmutableList<T> list) {
         checkNotNull(list, "list");
-        if (list.isEmpty()) { return this; }
```

Comment:
It's a duplicate check, especially since it cannot be removed from the `addRecursive()` method (as that is peeling off the `list` and needs the empty check to end recursion eventually). So if we keep it there we have duplicate code as well as a redundant duplicate check on all added non-empty lists.